Add object keypoint similarity method

getzze opened the pull request:
Using the flow tracker with two mice, I was sometimes getting unexpected identity switches when one of the mouse was missing half or more of its keypoints.
I figured it was a problem with the `instance_similarity` function. So I wrote a new `object_keypoint_similarity` function (in fact a function factory because it has parameters). The `instance_similarity` function compute the distance between each keypoints from a reference instance and a query instance, takes the `exp(-d**2)`, sum for all the keypoints and divide by the number of visible keypoints in the reference instance.

This is a description of the three changes I did and why:
1. Adding an scale to the distance between the reference and query keypoint. Otherwise, if the ref and query keypoints are 3 pixels apart, they contribute to 0.0001 to the similarity score, versus 0.36 if they are 1 pixel apart. This is very sensitive to single pixel fluctuations.
Instead, the distance is divided by a user-defined pixel scale before applying the gaussian function. The scale can be chosen to be the error for each keypoint found during training of the model with the validation set. Ideally this could be retrieved automatically, it is now hidden in the `metrics.val.npz` file of the model.
This is what they use in [this paper](https://arxiv.org/abs/1707.05388).

2. The prediction score for each keypoint can be used to weigh the influence of each keypoint similarity in the total similarity. Like this, uncertain keypoints will not bias the total similarity.

3. Dividing the sum of individual keypoint similarities by the number of visible keypoints in the reference instance results in higher similarity scores if the reference has few keypoints (meaning a bad reference instance). Imagine a query instance with 4 keypoints:
      -  a first ref instance with 1 keypoint that matches exactly one keypoint: similarity = exp(-0)/1 = 1
      -  a second ref instance with 4 keypoints, where 3 keypoints only match exactly the query instance: similarity = (1+1+1)/4 = 0.75
Dividing by the total number of keypoints instead gives 0.25 and 0.75 respectively, which is preferable.

I didn't create a cli option to change the point 3, but it can be easily added. Implementing points 1 and 3 dramatically improved the tracking.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
	- Enhanced tracking capabilities with new parameters for object keypoint similarity.
	- Added a new pytest fixture for sorted frame predictions to improve test usability.

- **Bug Fixes**
	- Improved handling of inference parameters for better prediction behavior.

- **Tests**
	- Updated tracking test functions to incorporate new similarity method parameters and improved data handling.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

roomrys reviewed (changes requested):
Sorry for the delay! This is a great new feature! OKS implementation looks good. A few suggestions to the pipeline form for display purposes. (Ideally we make this as a stacked option under the similarity method, but Qt is battling me - I'll mess with this a bit more).

getzze commented on sleap/nn/tracker/components.py:
```diff
@@ -40,6 +40,78 @@ def instance_similarity(
+        if max_n_keypoints == 0:
+            return 0
+
+        # Compute distances
+        if kp_precision.size > 1 and 2 * kp_precision.size != ref_points.size:
+            mess = (
+                "keypoint_errors array should have the same size as the number of "
+                f"keypoints in the instance: {kp_precision.size} != {ref_points.size // 2}"
+            )
+            raise ValueError(mess)
```
hey, thanks for reviewing. In retrospect, raising an error is a bit too much, using a good substitute would be better. I prefer the truncated/ones-padded `kp_precision` with an error message.

getzze commented:
I added the option to change `normalization_keypoints` from the CLI at least.

getzze commented on sleap/nn/tracking.py:
```diff
@@ -763,6 +775,30 @@ def int_list_func(s):
+        def float_list_func(s):
+            return [float(x.strip()) for x in s.split(",")] if s else None
+
+        option = dict(name="oks_errors", default="")
```
I tried several videos with different zoom values and I find 5 is a better default value. I only have videos with two mice, but it would be worth checking the `metrics.val.npz` of other animals and points of view, to see if it makes sense to put 5 as default instead of 1.
The code to get the error from the validation set on each keypoint:

```python
p = "metrics.val.npz"
a = np.load(p, allow_pickle=True)
ds = a["metrics"][()]["dist.dists"]
np.nanmean(ds, axis=0)
```

getzze commented on sleap/config/pipeline_form.yaml:
```diff
@@ -323,7 +339,7 @@ inference:
       type: list
       default: iou
-      options: instance,centroid,iou
+      options: instance,centroid,iou,object_keypoint
```
I left "object_keypoint" without a space, otherwise you have to use quotes when using this option from the CLI.

roomrys replied:
CLI arguments for similarity policies are directly pulled from the `similarity_policies` dictionary, so I added the space. No quotes needed when calling from CLI, but you will need the underscore.
https://github.com/talmolab/sleap/blob/eac2e2bc2e7a98f61dd8efc90540433da69e846f/sleap/nn/tracking.py#L752-L755
https://github.com/talmolab/sleap/blob/eac2e2bc2e7a98f61dd8efc90540433da69e846f/sleap/nn/tracking.py#L342-L346

getzze replied:
When running from the GUI, what is passed to the CLI is:
```
tracking.similarity = 'object keypoint'
```
(with space), which is not recognized as a valid similarity function.

The entry from the GUI should be formatted to replace the space by an underscore in `gui/learning/runners.py`. I will push a commit.

roomrys commented on tests/nn/test_tracker_components.py:
```diff
@@ -9,21 +9,57 @@
+            inst.track = None
+
+        track_args = dict(untracked_instances=lf.instances, img=lf.video[lf.frame_idx])
+        t.track(**track_args)
```
This test takes relatively long, use sparingly and with as few frames as possible.

roomrys commented on sleap/gui/learning/dialog.py:
```diff
@@ -770,6 +770,7 @@ def __init__(
         self, mode: Text, skeleton: Optional["Skeleton"] = None, *args, **kwargs
     ):
         super(TrainingPipelineWidget, self).__init__(*args, **kwargs)
+        self.setMinimumHeight(720)  # Hard-code minimum size due to layout problems
```
Flagging this because it's a shortcut to dealing with a jumpy GUI.

roomrys reviewed (commented):
I added tests to get line coverage, but haven't done any tests on performance.

getzze commented on sleap/nn/tracker/components.py:
```diff
@@ -71,7 +71,8 @@ def factory_object_keypoint_similarity(
     """
     keypoint_errors = 1 if keypoint_errors is None else keypoint_errors
-    kp_precision = 1 / (2 * np.array(keypoint_errors) ** 2)
+    with np.errstate(divide='ignore'):
```
If you want to ignore key points for the tracking, you can set the error to 0 for the specific key point.
This commit silences the "divide by zero" warnings, as this is an allowed operation.

talmo commented:
Hi folks, are we ready to merge this? Do we need to change the UI a bit more still?

getzze commented:
Hi, I reverted the GUI to the original proposal (without stacked widget) that was more convenient (or less annoying).
I think this is good to go now!

But maybe there should be a revamp of the tracking window (in another PR) because it has grown quite a bit with my PRs :D 
Maybe the Kalman filter part could go as it is not working great. Or the tracking options could be in another tab like the inference options...

roomrys commented:
Hi @getzze,

Yes, you are adding too many features for the GUI to handle! kudos 😎 The hold-up to merge this has indeed been displaying all the new features. I like your proposals for re-organizing the Training/Inference Pipeline dialog. Also agreed that those should be handled in a different PR.

I am going to hold-off on merging this until after our next release (I'd like both the Training dialog and the PR to be included in the same release, but the revamping won't be happening prior to the long over-do 1.3.0). Aiming to get 1.3.0 out by the end of this week, then will be working on much needed GUI revamping to accompany this PR.

Thanks!
Liezl

getzze commented:
Hey, I just wanted to bump this PR, as it is very useful (at least to me) so I would like to see you in the main branch. Thanks!

getzze commented on sleap/nn/tracking.py:
```diff
@@ -466,14 +468,15 @@
         tracks = []
         for track, matched_instances in track_matching_queue_dict.items():
-            if len(tracks) <= self.max_tracks:
+            if not max_tracking or len(tracks) <= self.max_tracks:
```
corrected

getzze commented:
Hi @roomrys  @talmo
I rebased this PR, I hope it will make it to version 1.4 (oldest open PR now !!).
The UI problem has been solved so nothing should block it.

Tests in `tests/nn/test_tracker_components.py` were not passing because of bugs introduced by the `max_tracks` new feature that were actually not tested. Now `max_track` is also tested and fixed.

Cheers

talmo reviewed (approved):
Thanks for following up here @getzze!! This looks great to me :)

getzze commented:
Thanks!
